app/models/comfy/cms/page.rb

```diff
@@ -3,14 +3,28 @@
 class Comfy::Cms::Page < ActiveRecord::Base
   scope :most_popular, -> (number_of_items) { unscoped.order('page_views desc').limit(number_of_items) }
 
-  def self.all_english_articles
+  scope :with_tags, -> (tags) { joins(:taggings).where('tag_id IN (?)', tags.map(&:id)) }
+
+  scope :remove_self_from_results, -> (page) { where.not(id: page.id) }
+
+  scope :remove_alternate_language_page_when_welsh, (lambda do |page|
+    where.not(id: page.mirrors.first.try(:id))
```

Comment:
!!! I just remembered we tried to do it this way on Wednesday and it didn't work. It's a really tough one to debug because `Page#mirrors` isn't available in `rails c` for some reason?

But the root of it is that when you run a query on your own ActiveRecord class in a scope, it executes the query _within all your chained scopes_. So when `#mirrors` queries `Page`, it'll tag all the existing criteria on and you won't get the right mirrors.

If you want to see for yourself, look at the SQL that is generated when you run `#related_links`:

```
Q1: Tag Load (1.1ms)  SELECT `tags`.* FROM `tags` INNER JOIN `taggings` ON `tags`.`id` = `taggings`.`tag_id` WHERE `taggings`.`taggable_id` = 605 AND `taggings`.`taggable_type` = 'Comfy::Cms::Page'

Q2: Comfy::Cms::Site Load (0.2ms)  SELECT `comfy_cms_sites`.* FROM `comfy_cms_sites`  WHERE `comfy_cms_sites`.`is_mirrored` = 1

Q3: Comfy::Cms::Page Load (0.4ms)  SELECT  `comfy_cms_pages`.* FROM `comfy_cms_pages` INNER JOIN `comfy_cms_layouts` ON `comfy_cms_layouts`.`id` = `comfy_cms_pages`.`layout_id` INNER JOIN `comfy_cms_sites` ON `comfy_cms_sites`.`id` = `comfy_cms_pages`.`site_id` INNER JOIN `taggings` ON `taggings`.`taggable_id` = `comfy_cms_pages`.`id` AND `taggings`.`taggable_type` = 'Comfy::Cms::Page' WHERE `comfy_cms_sites`.`label` = 'en' AND `comfy_cms_layouts`.`identifier` = 'article' AND `taggings`.`tag_id` IN (2, 10, 70) AND (`comfy_cms_pages`.`id` != 605) AND `comfy_cms_pages`.`site_id` = 2 AND `comfy_cms_pages`.`full_path` = '/do-you-need-to-borrow-money' GROUP BY comfy_cms_pages.id  ORDER BY COUNT(*) DESC, comfy_cms_pages.page_views DESC LIMIT 1

Q4: Comfy::Cms::Page Load (4.8ms)  SELECT  `comfy_cms_pages`.* FROM `comfy_cms_pages` INNER JOIN `comfy_cms_layouts` ON `comfy_cms_layouts`.`id` = `comfy_cms_pages`.`layout_id` INNER JOIN `comfy_cms_sites` ON `comfy_cms_sites`.`id` = `comfy_cms_pages`.`site_id` INNER JOIN `taggings` ON `taggings`.`taggable_id` = `comfy_cms_pages`.`id` AND `taggings`.`taggable_type` = 'Comfy::Cms::Page' WHERE `comfy_cms_sites`.`label` = 'en' AND `comfy_cms_layouts`.`identifier` = 'article' AND `taggings`.`tag_id` IN (2, 10, 70) AND (`comfy_cms_pages`.`id` != 605) AND (`comfy_cms_pages`.`id` IS NOT NULL) GROUP BY comfy_cms_pages.id  ORDER BY COUNT(*) DESC, comfy_cms_pages.page_views DESC LIMIT 3
```

Q3 fetches the mirrors, Q4 fetches the related links. If you scroll along you'll see they're virtually identical, except that right at the end of Q3 is the criteria for finding mirrors.